tests/test_api.py

```diff
@@ -0,0 +1,154 @@
+    for invalid_opt in mf.invalid_opts
+]
+
+
+@pytest.mark.parametrize("path,create_opts,invalid_arg", create_params_invalid)
+def test_create_invalid(path: str, create_opts: APIOpts, invalid_arg, client):
+    create_kwargs = create_opts.copy()
+    create_kwargs.update(invalid_arg)
+    with pytest.raises(client.error_cls):
+        _ = client.create(path, create_kwargs)
+
+
+@pytest.mark.parametrize("model_fixtures", ALL_MODEL_FIXTURES)
+def test_read_range(model_fixtures, client):
+    # first create some data
+    path = model_fixtures.path
+    for create_opts in model_fixtures.create_opts:
+        client.create(path, create_opts)
+    response = client.read_range(path)
+    for expected, actual in zip(model_fixtures.create_opts, response):
+        compare_response(expected, actual)
+        assert actual["id"] > 0
+
+
+@pytest.mark.parametrize("model_fixtures", ALL_MODEL_FIXTURES)
+def test_read_single(model_fixtures, client):
+    # first create some data
+    path = model_fixtures.path
+    for create_opts in model_fixtures.create_opts:
+        create_response = client.create(path, create_opts)
+        read_response = client.read_single(path, create_response["id"])
+        compare_response(create_opts, read_response)
+        assert read_response["id"] == create_response["id"]
+
+
+@pytest.mark.parametrize("model_fixtures", ALL_MODEL_FIXTURES)
+def test_read_nonexistent(model_fixtures, client):
+    # first create some data
+    path = model_fixtures.path
+    id = 99999999  # an id that we are pretty sure does not exist in the db!
+    with pytest.raises(client.error_cls):
+        _ = client.read_single(path, id)
+
+
+@pytest.mark.parametrize("path,create_opts,invalid_arg", create_params_invalid)
+def test_update_invalid(path: str, create_opts: APIOpts, invalid_arg, client):
+    response = client.create(path, create_opts)
+    id = response["id"]
+    with pytest.raises(client.error_cls):
+        _ = client.update(path, id, invalid_arg)
+
+
+update_params = [
+    (mf.path, create_opts, invalid_opt)
+    for mf in ALL_MODEL_FIXTURES
+    for create_opts in mf.create_opts[:1]  # just use the first create fixture
+    for invalid_opt in mf.update_opts
+]
+
+
+@pytest.mark.parametrize("path,create_opts,update_opts", update_params)
+def test_update(path: str, create_opts: APIOpts, update_opts, client):
+    create_response = client.create(path, create_opts)
+    id = create_response["id"]
+    response = client.update(path, id, update_opts)
+    compare_response(update_opts, response)
+
+
+@pytest.mark.parametrize("path,create_opts", create_params)
+def test_delete(path: str, create_opts: APIOpts, client):
+    data = client.create(path, create_opts)
+    id = data["id"]
+    _ = client.delete(path, id)
+    with pytest.raises(client.error_cls):
+        _ = client.read_single(path, id)
+
+
+@pytest.mark.parametrize("model_fixtures", ALL_MODEL_FIXTURES)
+def test_delete_invalid(model_fixtures, client):
+    id = 99999999
```

Comment:
Yep, this is a better way to handle this, rather than instantiating a model, as I had done previously.